Geometry/GEMGeometryBuilder/plugins/ME0GeometryValidate.cc

```diff
@@ -0,0 +1,341 @@
+  eventSetup.get<MuonGeometryRecord>().get(me0Geometry_);
+  if (me0Geometry_.isValid()) {
+    LogVerbatim("ME0Geometry") << "Validating ME0 chamber geometry";
+    validateME0ChamberGeometry();
+    validateME0EtaPartitionGeometry2();
+    validateME0EtaPartitionGeometry();
+  } else {
+    LogVerbatim("ME0Geometry") << "Invalid ME0 geometry";
+  }
+}
+
+void ME0GeometryValidate::validateME0ChamberGeometry() {
+  clearData();
+
+  for (auto const& it : me0Geometry_->chambers()) {
+    ME0DetId chId = it->id();
+    GlobalPoint gp = it->surface().toGlobal(LocalPoint(0.0, 0.0, 0.0));
+    const TGeoMatrix* matrix = fwGeometry_.getMatrix(chId.rawId());
+
+    if (!matrix) {
+      LogVerbatim("ME0Geometry") << "Failed to get matrix of ME0 chamber with detid: " << chId.rawId();
+      continue;
+    }
+    compareTransform(gp, matrix);
+
+    auto const& shape = fwGeometry_.getShapePars(chId.rawId());
+
+    if (!shape) {
+      LogVerbatim("ME0Geometry") << "Failed to get shape of ME0 chamber with detid: " << chId.rawId();
+      continue;
+    }
+    compareShape(it, shape);
+  }
+  makeHistograms("ME0 Chamber");
+}
+
+void ME0GeometryValidate::validateME0EtaPartitionGeometry2() {
+  clearData();
+
+  for (auto const& it : me0Geometry_->etaPartitions()) {
+    ME0DetId chId = it->id();
+    GlobalPoint gp = it->surface().toGlobal(LocalPoint(0.0, 0.0, 0.0));
+    const TGeoMatrix* matrix = fwGeometry_.getMatrix(chId.rawId());
+
+    if (!matrix) {
+      LogVerbatim("ME0Geometry") << "Failed to get matrix of ME0 eta partition 2 with detid: " << chId.rawId();
+      continue;
+    }
+    compareTransform(gp, matrix);
+
+    auto const& shape = fwGeometry_.getShapePars(chId.rawId());
+
+    if (!shape) {
+      LogVerbatim("ME0Geometry") << "Failed to get shape of ME0 eta partition 2 with detid: " << chId.rawId();
+      continue;
+    }
+    compareShape(it, shape);
+  }
+  makeHistograms("ME0 Eta Partition");
+}
+
+void ME0GeometryValidate::validateME0EtaPartitionGeometry() {
+  clearData2();
+
+  for (auto const& it : me0Geometry_->etaPartitions()) {
+    ME0DetId chId = it->id();
+    const int n_strips = it->nstrips();
+    const float n_pitch = it->pitch();
+    const StripTopology& topo = it->specificTopology();
+    const float stripLen = topo.stripLength();
+    const float* parameters = fwGeometry_.getParameters(chId.rawId());
+    nstrips_.push_back(fabs(n_strips - parameters[0]));
+
+    if (n_strips) {
+      for (int istrips = 1; istrips <= n_strips; istrips++) {
+        if (fabs(n_pitch - parameters[2]) < 1.0e-05)
+          pitch_.push_back(0.0f);
+        else
+          pitch_.push_back(fabs(n_pitch - parameters[2]));
+        if (fabs(stripLen - parameters[1]) < 1.0e-05)
+          pitch_.push_back(0.0f);
+        else
+          stripslen_.push_back(fabs(stripLen - parameters[1]));
+      }
+    } else {
+      LogVerbatim("ME0Geometry") << "ATTENTION! nStrips == 0";
```

Comment:
This message looks like it should be `LogWarning`.

Reply:
@cvuosalo : done